stumpy/mmotifs.py

```diff
@@ -0,0 +1,177 @@
+import numpy as np
+import logging
+
+# from .aamp_motifs import aamp_motifs
+from . import core, config, mdl, match
+
+logger = logging.getLogger(__name__)
+
+
+# @core.non_normalized(aamp_motifs)
+def mmotifs(
+    T,
+    P,
+    I,
+    min_neighbors=1,
+    max_distance=None,
+    cutoffs=None,
+    max_matches=10,
+    max_motifs=1,
+    atol=1e-8,
+    # normalize = True,
+    # p = 2.0
+):
+    """
+    Discover the top motifs for the multi-dimensional time series `T`
+
+    Parameters
+    ----------
+    T: numpy.ndarray
+        The multi-dimensional time series or sequence
+
+    P: numpy.ndarray
+        Multi-dimensional Matrix Profile of T
+
+    I: numpy.ndarray

```

Comment:
I'm trying to think if there's any possibility for us to not require `I` as an input and for us to recompute the nearest neighbor on-the-fly? Primarily, I'm asking because `motif()` does not require `I` as input and being able to remove it (and recomputing the parts that we need) in `mmotif()` would make it easy to have users call `motifs()` and, internally, have it figure out that the inputs are multi-dimensional and then pass the computation over to `mmotif()`. This is possible if `I` is not required as input. This would essentially require computing the distance profile for the candidate motif in order to find the nearest neighbor. However, if it adds a ton of noise to the nice code then it may not be worth it.

Reply:
In the multidimensional case, is it also certain that the motif pair has the same distance? In the one-dimensional case, the motif pair is obtained by finding the two lowest values of the same size (at least that's what the paper says). Couldn't we then simply pick out in `P` for each dimension the indices that have the same distance as the "candidate motif"? That we not only store the `candidate_idx`, but also `candidate_dist` and then look in `P` at which index places the distances are identical? Or am I just thinking too simply? Or does the same distances possibly only apply to the first motif?

Reply:
In both the 1-D case as well as the multi-dimensional case, I believe that the global motif (i.e., the global minimum) must occur as a pair. However, beyond the global minimum, the rest are **not** guaranteed to be a pair. To put it more concretely, let's say you have a time series with only 5 candidate motifs, namely `A`, `B`, `C`, `D`, and `E`, and `A` and `B` are the global motif pair, then `C` can have `D` as its one-nearest-neighbor but `D` can have `E` as its one-nearest-neighbor. And so, beyond the first candidate motif, the rest are never guaranteed to be a pair. So, to answer your question, it is not sufficient to only use `P` but, instead, we'd use `P` to choose the next candidate and then use that (set of) subsequence to compute the corresponding distance profile, extract the nearest neighbor index, and then proceed. Of course, we are repeating work due to the fact that we already have `I`. The more I think out loud, it's probably not a good idea to omit `I`. 

From what little that I've gathered, I've always interpreted a `motif` as the global minimum but I think many others have loosened that definition (which is fine).

Reply:
I think that could then also be quite computationally intensive if you have a large data set with many dimensions, right?

> users call motifs() and, internally, have it figure out that the inputs are multi-dimensional and then pass the computation over to mmotif()

Theoretically, however, it would still work if you make it clear in the description that an `I` must be passed to the function as soon as you work with multiple dimensions. Otherwise, the parameter can be set to `None` by default, so that in the one-dimensional case nothing happens with `I`. Or do you think that would be too confusing?

Reply:
Okay, let's keep `I`!